src/Processors/QueryPlan/ReadFromStorageStep.cpp

```diff
@@ -107,6 +107,18 @@ ReadFromStorageStep::ReadFromStorageStep(
             limits.speed_limits.min_execution_bps = settings.min_execution_speed_bytes;
         }
 
+        /**
+         * Leaf read limits should be applied only for local processing of distributed queries.
+         * Such limits allow to control the read stage on leaf nodes and exclude the merging stage.
+         * Consider the case when distributed query needs to read from multiple shards. Then leaf
+         * limits will be applied on the shards only (including the root node) but will be ignored
+         * on the results merging stage.
+         */
+        if (!storage->isRemote())
```

Comment:
The problem with this place is:

- `select * from remote('127.3', system.numbers) limit 100 settings max_rows_to_read_leaf=1` - will work
- `select * from remote('127.1', system.numbers) limit 100 settings max_rows_to_read_leaf=1` - will not

I though about different approaches to get a cleaner implementation, but I haven't come to anything yet

Reply:
Added test for this case